cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py

```diff
@@ -1989,3 +2012,78 @@ def keystone_config():
             create_self_config(ca, client)
         generate_keystone_configmap()
         set_state('keystone.credentials.configured')
+
+
+@when('vault.available',
+      'layer.vaultlocker.ready')
+@when_not('kubernetes-master.secure-storage.created')
+def create_secure_storage():
+    encryption_conf_dir = encryption_config_path().parent
+    encryption_conf_dir.mkdir(mode=0o700, parents=True, exist_ok=True)
+    try:
+        device_name = unitdata.kv().get('kubernetes-master.loop-device')
+        if not device_name:
+            check_call(['modprobe', 'loop'])
+            crypt_file = '/root/cdk/encrypted-config'
+            check_call(['dd', 'if=/dev/urandom', 'of={}'.format(crypt_file),
+                        'bs=1M', 'count=20'])
+            check_call(['losetup', '-f', crypt_file])
+            losetup_output = check_output(['losetup', '--json'])
+            device_name = [d['name']
+                           for d in json.loads(losetup_output)['loopdevices']
+                           if d['back-file'] == crypt_file][0]
+            unitdata.kv().set('kubernetes-master.loop-device', device_name)
+        vaultlocker.encrypt_device(device_name, str(encryption_conf_dir))
```

Comment:
Will this survive a reboot? I don't think it will -- I tried `losetup -f <file>` on my laptop and it didn't persist after reboot. I don't see any code calling it on reboot either.

The fstab entry (created [here](https://github.com/juju-solutions/layer-vaultlocker/blob/c31d7ad203d146d3b0b04e917235e73408a5c4ce/lib/charms/layer/vaultlocker.py#L85)) seems like it won't accomplish much, given that the loop device we want no longer exists after the reboot. I'd also worry about mounting some other device by mistake -- it seems unlikely, but if something else on the system is creating a loop device that happens to be formatted XFS, I could see it happening.

Reply:
Hrm.  I *thought* that dm-crypt or VaultLocker would take care of that, basing that mostly on [this](https://wiki.archlinux.org/index.php/Dm-crypt/Encrypting_a_non-root_file_system#Loop_device) not mentioning needing to setup a fstab entry, but I don't actually have any confirmation of that.  Perhaps @javacruft can chime in?

Reply:
After some testing yesterday, it seems that this is definitely a big issue.  Not only does the loop device not survive a reboot, but it not being there breaks the service entry created by VaultLocker which seems to cause the entire boot process to fail (at least, when rebooted, the instance didn't even come back up to a point where I could SSH in).  This means we need to ensure the loop device is created prior to the VaultLocker service running, and I'm not sure how to do that.

Reply:
This should be addressed by the latest commit and https://github.com/juju-solutions/layer-vaultlocker/pull/2